js/treetest.js

```diff
@@ -115,8 +130,80 @@ var tasks = {
 			var copyOfHistory = $.extend(true, [], this.answers[this.idx]);
 			expandToNode(copyOfHistory);
 		}
-	}
+		this.onSet.forEach(callback => callback(number));
+	},
+	onSet: [],
+	onSubmit: [],
 }
+
+var socket = {
+	_socket: null,
+
+	connect: function() {
+		this._socket = io();
+		this._socket.on('connect', () => {
+			console.debug('Socket connected');
+			this._socket.on('disconnect', () => {
+				console.debug('Socket disconnected');
+			})
+		})
+	},
+
+	emitPageLoadEvent: function() {
+		this._emitEvent({}, 'page load');
+	},
+
+	emitActivateNodeEvent: function(node) {
+		const data = {
+			node: $('#tree').jstree().get_path(node),
+			current_task_index: tasks.idx,
+		};
+		this._emitEvent(data, 'activate node');
+	},
+
+	emitOpenNodeEvent: function(node) {
+		const data = {
+			node: $('#tree').jstree().get_path(node),
+			current_task_index: tasks.idx,
+		};
+		this._emitEvent(data, 'open node');
+	},
+
+	emitCloseNodeEvent: function(node) {
+		const data = {
+			node: $('#tree').jstree().get_path(node),
+			current_task_index: tasks.idx,
+		};
+		this._emitEvent(data, 'close node');
+	},
+
+	emitTaskChangedEvent: function(newTaskIndex) {
+		this._emitEvent({ new_task_index: newTaskIndex }, 'task changed');
+	},
+
+	emitSubmitResponseEvent: function(answers) {
+		this._emitEvent({ answers }, 'submit response');
+	},
+
+	emitWindowVisibilityChangedEvent: function(newVisibilityState) {
+		const data = { new_visibility_state: newVisibilityState };
+		this._emitEvent(data, 'window visibility changed');
+	},
+
+	_emitEvent: function(data, eventType) {
+		const json = JSON.stringify({
+			id: crypto.randomUUID(),
+			timestamp: new Date().toISOString(),
+			tree_test_id: document.split('/').pop(),
+			...data,
+		});
+		const emitUntilAcknowledged = () => this._socket
+			.timeout(2000)
```

Comment:
Based on the documentation, any pending event will be lost if the user refreshes the tab
https://socket.io/docs/v4/delivery-guarantees

Reply:
Indeed. I think that is the limits of how much SocketIO does to guarantee delivery. I think if we want better guarantees, we will have to store events locally ourselves, e.g. with local storage or session storage, and using that storage to guarantee delivery would be a significant amount of extra complexity I think

Reply:
Does this covers the following AC?
```
Given that a user is connected via real-time communication, 
When the connection drops, 
Then the client temporarily stores the events locally and retries sending upon reconnection, ensuring no events are lost.
```

Reply:
I interpreted 'when the connection drops' to mean things like:

- The user's laptop gets disconnected from their WiFi network
- A random temporary network error in a data centre in Manchester somewhere prevents data getting through for a few seconds
- The user's phone loses 3G signal

These kinds of cases, SocketIO handles for free on the client.

But SocketIO doesn't use any persistent storage, so it can't cover cases like these:

- The user refreshes the tab
- The browser crashes
- The user kills the browser process
- The user drops their phone and it smashes

To handle these cases, we would need storage which persists across user sessions, and we would need to implement and manage this ourselves. I think that would be quite a lot of extra complexity

Reply:
That seems suitable to me. 👍